File: S6-ReinforcementLearning/Pt1-UCB/ucb.py
## Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


## Importing the dataset
dataset = pd.read_csv('Ads_CTR_Optimisation.csv')


## Implementing UCB
# Step 1: At each round n, we consider two numbers for easch ad i
N = 10000
d = 10
ads_selected = []
numbers_of_selecting = [0] * d   # << Ni(n)
sums_of_rewards = [0] * d        # << Ri(n)
total_reward = 0

# Step 2: From these two numbers we compute:
import math
for n in range(0, N):
    ad = 0
    max_upper_bound = 0
    
    for i in range(0, d):
        if numbers_of_selecting[i] > 0:
            # - The average reward of ad i up to round n
            average_reward = sums_of_rewards[i] / numbers_of_selecting[i]   # << ri(n)
            # - The Confidence Interval at round n
            delta_i = math.sqrt(3/2 * math.log(n) / numbers_of_selecting[i])    # << Δi(n)
            
            # Step 3: We select the ad i that has the maximum UCB    
            upper_bound = average_reward + delta_i
        else:
            upper_bound = 1e400
        
        if upper_bound > max_upper_bound:
            max_upper_bound = upper_bound
            ad = i

    ads_selected.append(ad)
    numbers_of_selecting[ad] += 1
    reward = dataset.values[n, ad]
    sums_of_rewards[ad] += reward
    total_reward += reward
            

## Visualizing the results
plt.hist(ads_selected)
plt.title('Histogram of Ads Selection')
plt.xlabel('Ads')
plt.ylabel('Number of times each ad was selected')
plt.show()

File: S6-ReinforcementLearning/Pt1-UCB/ucb.R
## Upper Confidence Bracket

## Importing the dataset
dataset = read.csv('Ads_CTR_Optimisation.csv')

## Implementing the UCB
# Step 1: At each round n, we consider two numbers for each ad i
N = 10000
d = 10
ads_selected = integer(0)
numbers_of_selection = integer(d)   # << Ni(n)
sums_of_the_reward = integer(d)     # << Ri(n)
total_reward = 0

# Step 2: From these two numbers we compute:
for (n in 1:N) {
  max_upper_bound = 0
  ad = 0
  
  for (i in 1:d) {
    if (numbers_of_selection[i] > 0) {
      # - The average reward of ad i up to round n
      average_reward = sums_of_the_reward[i] / numbers_of_selection[i]
      # - The confidence interval at round n
      delta_i = sqrt(3 / 2 * log(n) / numbers_of_selection[i])
      
      upper_bound = average_reward + delta_i
    } else {
      upper_bound = 1e400
    }
    
    # Step 3: We select the ad i that has the maximum UCB
    if (upper_bound > max_upper_bound) {
      max_upper_bound = upper_bound
      ad = i
    }
  }
  
  ads_selected = append(ads_selected, ad)
  numbers_of_selection[ad] = numbers_of_selection[ad] + 1
  reward = dataset[n, ad]
  sums_of_the_reward[ad] = sums_of_the_reward[ad] + reward
  total_reward = total_reward + reward
}

## Visualizing the results
hist(
  ads_selected, col = 'blue',
  main = 'Histogram of Ads Selection', 
  xlab = 'Ads', ylab = 'Number of times each ad was selected'
)
